[Merged by Bors] - Bloom

JMS55 opened the pull request:
# Objective

- Adds a bloom pass for HDR-enabled Camera3ds.
- Supersedes (and all credit due to!) https://github.com/bevyengine/bevy/pull/3430 and https://github.com/bevyengine/bevy/pull/2876

![image](https://user-images.githubusercontent.com/47158642/198698783-228edc00-20b5-4218-a613-331ccd474f38.png)

## Solution

- A threshold is applied to isolate emissive samples, and then a series of downscale and upscaling passes are applied and composited together.
- Bloom is applied to 2d or 3d Cameras with hdr: true and a BloomSettings component.

---

## Changelog

- Added a `core_pipeline::bloom::BloomSettings` component.
- Added `BloomNode` that runs between the main pass and tonemapping.
- Added a `BloomPlugin` that is loaded as part of CorePipelinePlugin.
- Added a bloom example project.

IceSentry commented on crates/bevy_pbr/src/bloom/bloom.wgsl:
```diff
@@ -0,0 +1,131 @@
+	scale: f32,
+};
+
+@group(0) @binding(0) var org: texture_2d<f32>;
```
Small nitpick, but most bevy shaders put the attributes on their own line. Would be nice if we had a wgslfmt to standardize formatting.

JMS55 replied:
Ngl I kinda hate the separate line thing, it bloats the file so much and makes it harder to read. But I can change it to match.

IceSentry replied:
Yeah, I don't really mind either way, I just want everything standardized automatically like all the rust file and never have to think about it.

IceSentry commented on crates/bevy_pbr/src/bloom/bloom.wgsl:
```diff
@@ -0,0 +1,131 @@
+    return color * max(rq, br - threshold) / max(br, 0.0001);
+}
+
+// samples org around the supplied uv using a filter
```
I'm not sure what you mean by "samples org". What does org mean here?

JMS55 replied:
I think original? Naming, especially in the shader, is something I need to clean up. I didn't write this code.

IceSentry replied:
Ah, right, I forgot this was based on another PR.

superdump replied:
'organised' perhaps?

IceSentry commented on crates/bevy_pbr/src/bloom/bloom.wgsl:
```diff
@@ -0,0 +1,131 @@
+// to it looking better
+//
+// these advantages are outlined in a youtube video by the Cherno:
+//   https://www.youtube.com/watch?v=tI70-HIc5ro
```
Is this the only reference you used? It's a great reference, but just curious if you used anything else.

JMS55 replied:
I didn't write this code. I think “Next Generation Post Processing in Call of Duty: Advanced Warfare” is also a good reference iirc.

IceSentry commented on crates/bevy_pbr/src/bloom/mod.rs:
```diff
@@ -0,0 +1,769 @@
+const BLOOM_SHADER_HANDLE: HandleUntyped =
+    HandleUntyped::weak_from_u64(Shader::TYPE_UUID, 929599476923908);
+
+pub struct BloomPlugin;
```
Am I correct in thinking that this should also work in 2d? If that's the case, I wonder if we could share the plugin for 2d and 3d. It might make sense to move this to bevy_core_pipeline if that's the case and extract the node to it's own file. A bit like the tonemapping and upscaling.

JMS55 replied:
Oh, hmm I didn't think of 2d. It should work in 2d I think, I would just need to remove the Camera3d filter. 

Should we make a bevy_post_process crate or something similar? I'm not sure if AA, tonemapping, upscaling, bloom, etc belong in either bevy_pbr/bevy_sprite or bevy_core_pipeline.

IceSentry replied:
I think, for this PR, keep it in core_pipeline, but we should probably have a new crate in the future.

DGriffin91 replied:
I think it would also need to be added to the `bevy_core_pipeline::core_2d` graph.

superdump replied:
Did we decide not to do 2D? Why?

IceSentry reviewed (commented):
First review pass looked good, but I didn't have time to go really deep.

JMS55 commented:
Should we add the BloomPlugin by default as part of PbrPlugin btw? It's something I carried over from the previous PR, unsure how I feel about it.

cart commented:
I think the bloom plugin should be added by default, but disabled by default. The functionality should be compiled in, but it should have zero cost when disabled.

JMS55 commented:
Ok, that's kinda what we have now. Note that it won't be _entirely_ 0 cost, as there's still the plugin registration and pipeline creation at startup, and probably the node running (and then immediately returning).

Something else I've realized: I'm not sure I'm handling viewports correctly. I need to test this on one of the examples that have different viewport vs window/texture sizes.

superdump commented on examples/3d/bloom.rs:
```diff
@@ -0,0 +1,81 @@
+
+fn main() {
+    App::new()
+        .insert_resource(Msaa { samples: 1 }) // 1. MSAA must be turned off
```
Why?

cart replied:
This is no longer true. I fixed it a few commits up. We can remove this line.

superdump commented on examples/3d/bloom.rs:
```diff
@@ -0,0 +1,81 @@
+        for z in -10..10 {
+            let mut hasher = DefaultHasher::new();
+            (x, z).hash(&mut hasher);
+            let rand = hasher.finish() % 2 == 0;
+
+            let material = match rand {
+                true => material.clone(),
+                false => material_non_emissive.clone(),
+            };
```
I'm not a fan of matching on booleans. That's what if else is for imo. :)

cart replied:
Agreed!

superdump commented on crates/bevy_pbr/src/bloom/bloom.wgsl:
```diff
@@ -0,0 +1,132 @@
+#import bevy_core_pipeline::fullscreen_vertex_shader
+
+struct Uniforms {
+	threshold: f32,
+	knee: f32,
+	scale: f32,
```
Tabs?
```suggestion
    threshold: f32,
    knee: f32,
    scale: f32,
```

superdump commented on crates/bevy_pbr/src/bloom/bloom.wgsl:
```diff
@@ -0,0 +1,132 @@
+// o   o   o
+//
+// this is used because it has a number of advantages that
+// outway the cost of 13 samples that basically boil down
```
```suggestion
// outweigh the cost of 13 samples that basically boil down
```

superdump commented on crates/bevy_pbr/src/bloom/mod.rs:
```diff
@@ -0,0 +1,760 @@
+                down_sampling_bind_groups.push(bind_group);
+            }
+
+            let mut up_sampling_bind_groups = Vec::new(); // TODO: Make these boxed slices
```
Why the todo? It could just be `Vec::with_capacity()`?

JMS55 replied:
It could, but it's slightly more expensive (1 extra usize, iirc) to hold a Vec when we know the capacity won't change.

JMS55 replied:
Yep, good point.

superdump reviewed (changes requested):
Some nits. Please also swap all `down_sampl` with `downsampl` and `up_sampl` with `upsampl` and `pre_filter` with `prefilter`.

cart commented on examples/3d/bloom.rs:
```diff
@@ -0,0 +1,81 @@
+    commands.spawn((
+        Camera3dBundle {
+            camera: Camera {
+                hdr: true, // 2. HDR must be enabled on the camera
```
No bloom.

cart replied:
(graceful failure)

JMS55 commented:
Addressed all the feedback. All that's missing at this point is:
1. Improving BloomSettings parameter documentation
2. Getting the bloom example to run

I tried adding some UI/controls to the bloom example, but I'm getting the following error. Help would be greatly appreciated!
```
Caused by:
    In a RenderPass
      note: encoder = `<CommandBuffer-(0, 1, Vulkan)>`
    In a set_pipeline command
      note: render pipeline = `ui_pipeline`
    Render pipeline targets are incompatible with render pass
    Incompatible color attachment: the renderpass expected [Some(Rgba16Float)] but was given [Some(Rgba8UnormSrgb)]
```

JMS55 commented:
Using renderdoc to inspect the results, I've also noticed that the final downsample/mip seems to always be an empty black texture.

cart commented:
> I tried adding some UI/controls to the bloom example, but I'm getting the following error. Help would be greatly appreciated!

Looks like we need to specialize the UI pipeline on "hdrness" so the color attachment format matches. Apparently I forgot to test this case in #3425

cart commented:
(fixing that now)

cart commented on examples/3d/bloom.rs:
```diff
@@ -0,0 +1,168 @@
+    text.push_str("D/F - Intensity\n");
+
+    if keycode.pressed(KeyCode::Q) {
+        bloom_settings.threshold -= 1.0;
```
All of these settings should be scaled with `time.delta_seconds()` if we're going to update them every frame a key is pressed.

cart commented on crates/bevy_core_pipeline/src/bloom/mod.rs:
```diff
@@ -0,0 +1,798 @@
+    view::ViewTarget,
+    Extract, RenderApp, RenderStage,
+};
+use bevy_utils::HashMap;
```
We need to import info_span when tracing is enabled.

```suggestion
#[cfg(feature = "trace")]
use bevy_utils::tracing::info_span;
use bevy_utils::HashMap;
```

cart reviewed (approved):
Once the build is fixed I think this is good to go.

cart commented:
bors r+

cart commented on Cargo.toml:
```diff
@@ -313,6 +313,10 @@ description = "Illustrates spot lights"
+[[example]]
+name = "bloom"
+path = "examples/3d/bloom.rs"
+
```
```suggestion

[package.metadata.example.bloom]
name = "Bloom"
description = "Illustrates bloom configuration using HDR and emissive materials"
category = "3D Rendering"
wasm = true
```

cart commented on examples/3d/bloom.rs:
```diff
@@ -0,0 +1,172 @@
+use bevy::{core_pipeline::bloom::BloomSettings, prelude::*};
```
```suggestion
//! Illustrates bloom configuration using HDR and emissive materials.

use bevy::{core_pipeline::bloom::BloomSettings, prelude::*};
```

cart commented:
bors r+

cart commented:
bors r-

cart commented:
Just realized we can't enable the bloom example on WASM because it multisamples by default (and hdr multisampling isn't supported on wasm)

cart commented on Cargo.toml:
```diff
@@ -313,6 +313,15 @@ description = "Illustrates spot lights"
+name = "Bloom"
+description = "Illustrates bloom configuration using HDR and emissive materials"
+category = "3D Rendering"
+wasm = true
```
```suggestion
```

cart commented:
bors r+

mockersf commented:
bors retry
